### PR TITLE
update wasmd to v0.51.0

### PR DESCRIPTION
**Pull request by kwtalley**
## Description
This PR updates wasmd to the latest [v0.51.0-pio-1](https://github.com/provenance-io/wasmd/tree/v0.51.0-pio-1) to support CosmWasm 2.0

closes: #2044 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

## Summary by CodeRabbit

- **New Features**
  - Added support for cosmwasm_2_0 feature in app initialization.
  - Introduced new RPC method `BuildAddress` for contract address generation.

- **Bug Fixes**
  - Resolved lockfile contention in app tests by creating a new temporary directory for `wasmvm` data.

- **Chores**
  - Updated import paths to use `wasmvm/v2` across multiple files.
  - Updated Dockerfile paths for wasmvm libraries to new version directories.
  - Overridden `WASMVM_REPO` in GitHub workflow to correct path.

- **Documentation**
  - Updated CHANGELOG.md with new versions of `wasmd` and `wasmvm` and added information on stargate queries and a new circuit breaker module.

### CODE REVIEW

**Comment by Taztingo** [comment]
Minor fix that should resolve the changelog error. Move the following to Dependencies, and use pull request number.


* Bump wasmd to `v0.51.0` [#2045](https://github.com/provenance-io/provenance/pull/2045).
* Bump wasmvm to `v2.0.1` [#2045](https://github.com/provenance-io/provenance/pull/2045).

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -62,10 +62,12 @@ jobs:
             make install
           binaries: |
             - /go/bin/provenanced
+# override WASMVM_REPO since github.com/CosmWasm/wasmvs/v2 path doesn't exist in github
```
Comment (iramiller) [inline .github/workflows/docker.yml:65]: While just a comment the spelling issue here: `wasmvs/v2` bugs me

Comment (kwtalley) [reply .github/workflows/docker.yml]: :cry: it actually wasn't a variable for the command so it's been reverted anyways

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -62,10 +62,12 @@ jobs:
...
           build-env: |
             - "WITH_LEDGER=false"
             - "WITH_CLEVELDB=false"
             - "BUILD_TAGS=muslc musl dynamic"
+            - "WASMVM_REPO=github.com/CosmWasm/wasmvm"
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply .github/workflows/docker.yml]: reverted commit

File: docker/blockchain/Dockerfile [hunk 3]
```diff
@@ -40,15 +40,15 @@ RUN apt-get update && \
...
+COPY --from=build /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.*.so /tmp
 COPY --from=build /go/bin/provenanced /usr/bin/provenanced
 
 RUN ARCH=$(uname -m) && \
     if [ "$ARCH" != "x86_64" ] && [ "$ARCH" != "aarch64" ]; then \
         echo "Unsupported architecture (required: x86_64 or aarch64): $ARCH"; \
         exit 1; \
     fi && \
-    cp /tmp/libwasmvm.$ARCH.so /usr/local/lib/. && \
+    cp /"tmp/libwasmvm.$ARCH.so /usr/local/lib/." && \
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply docker/blockchain/Dockerfile]: fixed

File: docker/blockchain/Dockerfile [hunk 4]
```diff
@@ -40,7 +40,7 @@ RUN apt-get update && \
     rm -rf /var/lib/apt/lists/
     
 
-COPY --from=build /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/internal/api/libwasmvm.*.so /tmp
+COPY --from=build /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.*.so /tmp
 COPY --from=build /go/bin/provenanced /usr/bin/provenanced
 
 RUN ARCH=$(uname -m) && \
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply docker/blockchain/Dockerfile:46]: fixed

File: docker/blockchain/Dockerfile [hunk 5]
```diff
@@ -40,7 +40,7 @@ RUN apt-get update && \
     rm -rf /var/lib/apt/lists/
     
 
-COPY --from=build /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/internal/api/libwasmvm.*.so /tmp
+COPY --from=build /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.*.so /tmp
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply docker/blockchain/Dockerfile:43]: fixed

File: networks/ibc/blockchain-ibc/Dockerfile [hunk 6]
```diff
@@ -36,7 +36,7 @@ WORKDIR /provenance
 EXPOSE 26656 26657 1317 9090
 
 # Source binaries from the build above
-COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/internal/api/libwasmvm.x86_64.so /usr/local/lib/
+COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.x86_64.so /usr/local/lib/
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply networks/ibc/blockchain-ibc/Dockerfile:39]: fixed

File: networks/local/blockchain-local/Dockerfile [hunk 7]
```diff
@@ -36,7 +36,7 @@ WORKDIR /provenance
 EXPOSE 26656 26657 1317 9090
 
 # Source binaries from the build above
-COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/internal/api/libwasmvm.x86_64.so /usr/local/lib/
+COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.x86_64.so /usr/local/lib/
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply networks/local/blockchain-local/Dockerfile:39]: fixed

File: networks/dev/blockchain-dev/Dockerfile [hunk 8]
```diff
@@ -35,7 +35,7 @@ WORKDIR /provenance
 EXPOSE 26656 26657 1317 9090
 
 # Source binaries from the build above
-COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/internal/api/libwasmvm.x86_64.so /usr/local/lib/
+COPY --from=build-x86_64 /go/src/github.com/provenance-io/provenance/vendor/github.com/CosmWasm/wasmvm/v2/internal/api/libwasmvm.x86_64.so /usr/local/lib/
```
[the remark this answers is not in the document]
Comment (kwtalley) [reply networks/dev/blockchain-dev/Dockerfile:38]: fixed

**Review by Taztingo — APPROVED** [review]

**Review by iramiller — APPROVED** [review]

**Pull request merged**